File: Trees/LC_297.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.30

class Codec {
public:
    // learn to use ostringstream and istringstream to represent the string
    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        ostringstream out;
        encode(root, out);
        return out.str();
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        istringstream in(data);
        return decode(in);
    }

private:
    void encode(TreeNode* root, ostringstream& out){
        if (root == nullptr){
            out << "# ";
            return;
        }
        out << root->val << " ";
        encode(root->left, out);
        encode(root->right, out); 
    }
    TreeNode* decode(istringstream& in){
        string str;
        in >> str;
        if (str == "#"){
            return nullptr;
        }
        TreeNode* root = new TreeNode(stoi(str));
        root->left = decode(in);
        root->right = decode(in);
        return root;
    }
};

File: Trees/LC_297.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.30

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # use the preorder traversal to encode!
        def encode(root, ans):
            if not root:
                ans += "# "
                return ans
            val = root.val
            ans += str(val) + " "
            ans = encode(root.left, ans)
            ans = encode(root.right, ans)
            return ans
        return encode(root, "")

    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        data_split = data.split(" ")
        
        def decode(data):
            if data[0] == "#":
                data.pop(0)
                return None
            val = int(data[0])
            data.pop(0)
            root = TreeNode(val)
            root.left = decode(data)
            root.right = decode(data)
            return root
        return decode(data_split)